workshop1/slides/src/slides.md

```diff
@@ -0,0 +1,664 @@
+
+makeCandidate :: [String] -> Maybe Candidate
+makeCandidate [c, p, s, a] =
+  fromMaybeAge (Candidate c p s) (readMaybe a)
+makeCandidate _ = Nothing
+
+fromMaybeAge
+  :: (Double -> Candidate)
+  -> Maybe Double
+  -> Maybe Candidate
+fromMaybeAge partialCandidate Nothing  = Nothing
+fromMaybeAge partialCandidate (Just a) = Just (partialCandidate a)
+```
+
+
+## Stripping Quotes
+
+```haskell
+stripQ :: String -> String
+stripQ = undefined
+  where
+    leftStrip :: String -> String
+    leftStrip ('"' : xs) = xs
+    leftStrip xs         = xs
+
+    rightStrip :: String -> String
+    rightStrip = undefined
+```
+
+* Use `leftStrip` and `reverse` to implement `rightStrip`
+* Use `leftStrip` and `rightStrip` to implement `stripQ`
+* `reverse` reverses items in a list
+* `where` used to define local functions
+
+## Stripping Quotes
+
+Our solution:
+```haskell
+stripQ :: String -> String
+stripQ = leftStrip . rightStrip
+  where
+    leftStrip :: String -> String
+    leftStrip ('"' : xs) = xs
+    leftStrip xs         = xs
+
+    rightStrip :: String -> String
+    rightStrip = reverse . leftStrip . reverse
+```
+
+* Point-free style: composition FTW!
+
+## Improving `parseLine`
+
+* Using `stripQ` to improve `makeCandidate`
+
+```haskell
+...
+
+makeCandidate :: [String] -> Maybe Candidate
+makeCandidate [c, p, s, a] =
+  let c' = stripQ c
+      p' = stripQ p
+      s' = stripQ s
+  in fromMaybeAge (Candidate c' p' s') (readMaybe a)
+makeCandidate _ = Nothing
+
+...
+```
+
+* `let ... in ...` binding useful to split up functions
+
+
+## The `map` function
+
+* How do we apply `parseLine` to every row?
+* Use `map`, a *higher-order function*
+  * Takes two inputs: a function and a list.
+  * Applies the function to every item in the list
+
+```haskell
+map :: (a -> b) -> [a] -> [b]
+map f []     = []
+map f (x:xs) = f x : map f xs
+```
+
+* Defined using *recursion*
+
+
+## Writing `parseFile :: CSV -> DataModel`
+
+* We are ready to write the full parser function:
+
+```haskell
+type CSV = String
+type DataModel = [Maybe Candidate]
+
+parseFile :: CSV -> DataModel
+parseFile = undefined
+```
+
+* `lines` splits strings into lines
+* `tail` removes the first item in a list
+* Try and use point-free style
+
+## Writing `parseFile :: CSV -> DataModel`
+
+Our solution:
+
+```haskell
+parseFile :: CSV -> DataModel
+parseFile = map parseLine . tail . lines
+```
+
+
+## Generating `encode :: DataModel -> JSON`
+
+```haskell
+{-# LANGUAGE DeriveGeneric #-}
+main Main where
+
+import           GHC.Generics
+import qualified Data.Aeson      as JSON
+import qualified Data.ByteString as B
+
+data Candidate = Candidate
+  { ...
+  } deriving (Show, Generic)
+
+instance Aeson.ToJSON Candidate
+```
+
+* Language extensions expand the Haskell language
+* `Data.Aeson` provides encoding to and from JSON
+* `Data.ByteString` provides efficient binary strings
+
+## Generating `encode :: DataModel -> JSON`
+
+```haskell
+type JSON = ByteString
```

Comment:
Wut? This is confusing, since we're using `JSON` as a module name already. Better changes one or the other to something else: what about `JSONValue`?

Reply:
I agree that it's confusing. JSONValue is also misleading because Aeson has a Value datatype, which is not a string. Maybe JSONString?

Reply:
`type` is a big source of confusion in compiler messages (servant being a case in point). Why not just use ByteString?

Reply:
I think both options would be better than the current state of affairs

Reply:
@chreekat servant is another different beast. I think the primary use case is differentiating `String`which is a single line of CSV vs a `String` which is the full file contents. I'm not confident on introducing newtypes at this point.

As for this specific location, I think it's just nic trying to be coherent on his type aliases.

Reply:
Changed to `type JSONByteString = ByteString`